store/src/java/com/zimbra/cs/service/PreviewServlet.java

```diff
@@ -0,0 +1,559 @@
+   * @param reason message string for {@link HttpServletResponse}
+   */
+  void respondWithError(HttpServletResponse resp, int errCode, String reason) {
+    resp.setContentType("text/html; charset=UTF-8");
+    try {
+      resp.sendError(errCode, reason);
+    } catch (IOException e) {
+      LOG.error(e.getMessage(), e);
+    }
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+      throws IOException, ServletException {
+    ZimbraLog.clearContext();
+    addRemoteIpToLoggingContext(req);
+
+    final AuthToken authToken = getAuthTokenFromCookie(req, resp);
+    checkAuthTokenFromCookieOrRespondWithError(authToken, req, resp);
+
+     final Pattern requiredQueryParametersPattern =
+        Pattern.compile(SERVLET_PATH + "/([a-zA-Z]+)/([0-9]+)/([0-9]+)");
+    final Matcher requiredQueryParametersMatcher =
+        requiredQueryParametersPattern.matcher(getUrlWithQueryParams(req));
+
+    // check url for the presence of required parameters and query string
+    // send error otherwise
+    if (req.getQueryString() == null || (!requiredQueryParametersMatcher.find()
+        || requiredQueryParametersMatcher.groupCount() != 3)) {
+      respondWithError(resp, HttpServletResponse.SC_BAD_REQUEST, "Missing required parameters");
+    } else {
+      final int messageId = Integer.parseInt(requiredQueryParametersMatcher.group(2));
+      final String partNo = requiredQueryParametersMatcher.group(3);
+
+      // get attachment
+      final Try<MimePart> attachmentMimePart = getAttachment(authToken, messageId, partNo);
+      if (attachmentMimePart.isFailure()) {
+        respondWithError(
+            resp, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, attachmentMimePart.failed().get().getMessage());
+      }
+
+      // get preview
+      final Try<BlobResponseStore> previewOfAttachment =
+          getAttachmentPreview(getUrlWithQueryParams(req), attachmentMimePart.get());
+      if (previewOfAttachment.isFailure()) {
+        if (previewOfAttachment.failed().get() instanceof BadRequest) {
+          respondWithError(
+              resp,
+              HttpServletResponse.SC_BAD_REQUEST,
+              previewOfAttachment.failed().get().getMessage());
+        } else if (previewOfAttachment.failed().get() instanceof ItemNotFound) {
+          respondWithError(
+              resp,
+              HttpServletResponse.SC_NOT_FOUND,
+              previewOfAttachment.failed().get().getMessage());
+        } else if (previewOfAttachment.failed().get() instanceof ValidationError) {
+          respondWithError(
+              resp,
+              HttpServletResponse.SC_UNAUTHORIZED,
+              previewOfAttachment.failed().get().getMessage());
+        } else {
+          respondWithError(
+              resp,
+              HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+              previewOfAttachment.failed().get().getMessage());
+        }
```

Comment:
Please use `.onFailure()` method instead of this `if-else` waterfall. In particular, there is this signature you can use: 
```java
onFailure(Class<X> exceptionType, Consumer<? super X> action)
```

Where you can specify the type of the exception and pass a lambda function that will contain the desired exception. So this becomes:
```suggestion
      final Try<BlobResponseStore> previewOfAttachment =      getAttachmentPreview(getUrlWithQueryParams(req), attachmentMimePart.get())
      .onFailure(BadRequest.class, ex -> respondWithError(resp, HttpServletResponse.SC_BAD_REQUEST, previewOfAttachment.getMessage())
      .onFailure( // Continue from there :))
```

Reply:
The idea was to not have a single chain doing more than one job.
I changed the code according to your suggestion, now we have a statement doing more than one task. Handling errors of preview service and committing the response.